homeassistant/components/solax/sensor.py

```diff
@@ -34,10 +49,28 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
     async_track_time_interval(hass, endpoint.async_refresh, SCAN_INTERVAL)
     devices = []
     for sensor, (idx, unit) in api.inverter.sensor_map().items():
```

Comment:
Ideally this is defined in Home Assistant (and not loaded from the library)

<https://github.com/squishykid/solax/blob/master/solax/inverter.py#L136-L169>

Having those in Home Assistant, allows for use of `EntityDescription`, which can have all the additional meta data needed (like `device_class`, `state_class` and such), which removes these kinda weird if/elif/else trees.

(For a future PR)

Reply:
Ack and @squishykid FYI

The tricky part is that the set of available sensors varies between inverter models, and the library is the canonical registry of those...

Reply:
We could have some base sensors defined in Home Assistant, and then add the rest based on the inverter model? They should all be able to produce the basics like: total PV power, grid power etc. some of them will have one or more phases of power though, so we can't determine that ahead of time from home assistant. In other cases the inverter might support a battery, but it won't be installed.

Reply:
> The tricky part is that the set of available sensors varies between inverter models, and the library is the canonical registry of those

Its not about defining them in 1 place, but 2 to decouple them. Home Assistant would have a list with sensors and their descriptions. The library would provide the sensors the matching sensors with that list and the list in Home Assistant gets loaded. That is the general pattern found across our codebase.

This reduces less "magic" code and from the code it is clear what the integration "could" provide. It also allows maintainers to search for certain usages inside codebase when e.g., an energy migration is happening 😉